### PR TITLE
enhancement(vrl): add `filter_array`

### PR DESCRIPTION
**Pull request by prognant**
Allow to remove elements from an array based on a regexp.
Non-byte type are removed.

Note: as this PR was mostly anticipating a usecase (datadog tags manipulation) I suggest we don't merge it unless there is a real need for the usecase **and** enumeration/iteration is not available in VRL.

### CODE REVIEW

**Comment by JeanMertz** [comment]
I haven't reviewed the PR itself yet, but I did want to note that this gets close to a "wait until we have iteration" use-case (#6031).

That is to say, with iteration, presumably, filtering would be a matter of iterating and then filtering individual fields in an array or object, which would make a function like this obsolete.

I'm not against merging this, but I do think we need some form of iteration sooner or later, at which point we'll likely deprecate functions that like this one.

**Comment by jszwedko** [comment]
Very much agreed that this seems like a great case for iteration. I'm also ok with merging this if it is blocking other work, but it will (hopefully!) be deprecated shortly.

File: docs/reference/remap/functions/filter_array.cue [hunk 1]
```diff
@@ -0,0 +1,50 @@
+package metadata
+
+remap: functions: filter_array: {
+	category: "Array"
+	description: """
+		Filters elements from the `value` array. It returns an array containint the elements matching the `pattern`.
```
Comment (pablosichert) [inline docs/reference/remap/functions/filter_array.cue:6]:
```suggestion
		Filters elements from the `value` array. It returns an array containing the elements matching the `pattern`.
```

File: lib/vrl/stdlib/src/filter_array.rs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        let pattern = arguments.required("pattern");
+
+        Ok(Box::new(FilterArrayFn { value, pattern }))
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::ARRAY,
```
Comment (pablosichert) [inline lib/vrl/stdlib/src/filter_array.rs:37]: I'm wondering if we can restrict the input type further to only accept arrays of `Value::Bytes`?

File: lib/vrl/stdlib/src/filter_array.rs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+impl Function for FilterArray {
+    fn identifier(&self) -> &'static str {
+        "filter_array"
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "some item match",
+                source: r#"filter_array(["foobar", "bazqux"], r'foo')"#,
```
Comment (pablosichert) [inline lib/vrl/stdlib/src/filter_array.rs:15]: I'm not sure if I would expect this to match if the regex doesn't match exhaustively. Maybe we can add an option for that behavior?

**Review by pablosichert — APPROVED** [review]
Looks good, @prognant!

Agree with @JeanMertz and @jszwedko regarding iteration.

Does this PR have a matching issue? If so, would be great to add it in the PR description as "Closes #XXX".

File: lib/vrl/stdlib/src/filter_array.rs (hunk 3 above)
Comment (leebenson) [reply lib/vrl/stdlib/src/filter_array.rs:15]: `r'foo'` matches _foobar_, so I'd expect the current behaviour.

Comment (JeanMertz) [reply lib/vrl/stdlib/src/filter_array.rs:15]: I think this is fine. We don't provide an exhaustive option elsewhere, and you can always use `^foo$` if need be.

**Comment by prognant** [comment]
> Does this PR have a matching issue? If so, would be great to add it in the PR description as "Closes #XXX".
Not really,  that work is related to easily enable datadog tags manipulation in VRL, e.g.:

```
tags = split(.ddtags)
tags = filter_array(tags, r'(beta|staging)')
tags = push("via:vector", tags)
.ddtags = join(tags, ",")

```

That being said I agree that adding enumeration/looping/mapping would make this PR essentially useless.

File: lib/vrl/stdlib/src/filter_array.rs (hunk 2 above)
Comment (prognant) [reply lib/vrl/stdlib/src/filter_array.rs:37]: I used the the same input kind as `match_array`,  I don't think we can enforce that kind of restriction here, I guess we could do that in `resolve` but after having a quick look to existing function I did not find any previous similar implementation.

**Comment by binarylogic** [comment]
@prognant what's the status here? Should we close since it seems to be stale?

**Comment by prognant** [comment]
> @prognant what's the status here? Should we close since it seems to be stale?

This was a provision in case this was needed, as VRL iteration is planned, this is irrelevant.

**Pull request closed without merging**